### PR TITLE
fix cmake bug for OpenCL

### PR DESCRIPTION
**Pull request by ffk0716**
*Description*: cmake failed with OpenCL build
*Testing*:
*Documentation*:
[Optional Fixes #issue]

Please see a detailed explanation of how to fill out the fields in the relevant sections in PULL_REQUEST.md.

### CODE REVIEW

**Comment by nadavrot** [comment]
Thank you for doing this work @ffk0716 !

**Comment by nadavrot** [comment]
@SplitInfinity

File: lib/Backends/OpenCL/CMakeLists.txt [hunk 1]
```diff
@@ -1,30 +1,38 @@
...
 
 add_custom_command(
-  OUTPUT "${CMAKE_BINARY_DIR}/glow/OpenCL/kernels_fwd_quantized_conv.cl.inc"
+  OUTPUT "${OpenCL_DIR}/kernels_fwd_quantized_conv.cl.inc"
   COMMAND include-bin
           "${CMAKE_CURRENT_SOURCE_DIR}/kernels_fwd_quantized_conv.cl"
-          "${CMAKE_BINARY_DIR}/glow/OpenCL/kernels_fwd_quantized_conv.cl.inc"
-  DEPENDS include-bin CPURuntime
+          "${OpenCL_DIR}/kernels_fwd_quantized_conv.cl.inc"
+  DEPENDS make_opencl_dir include-bin CPURuntime
```
Comment (opti-mix) [inline lib/Backends/OpenCL/CMakeLists.txt]: Not a problem in this PR, but why do we depend on `CPURuntime` in our `OpenCL` CMake file? Is it a copy&paste bug or something like this?

@rdzhabarov @SplitInfinity

**Comment by arunm-git** [comment]
CC: @SplitInfinity since the original changes to the opencl cmakelists.txt file were made to handle internal builds.
If the location of the .h files is changing through this PR, that will need to be handled internally too.

**Comment by SplitInfinity** [comment]
@arunm-git 

I made this change because handling it internally is harder (cc: @bertmaher). In #2721 I relied only on CI which was a mistake, so let me revisit that and figure out how to move the kernels properly without breaking the build.

**Comment by SplitInfinity** [comment]
Oh, the location of the files is not changing, never mind.

File: lib/Backends/OpenCL/CMakeLists.txt (hunk 1 above)
Comment (SplitInfinity) [reply lib/Backends/OpenCL/CMakeLists.txt]: Hmm, strange, I'm not sure...

**Comment by SplitInfinity** [comment]
Is anyone else able to reproduce this?

I just built from trunk and got:
```
meghanl-mbp:build_DEBUG meghanl$ ./tests/OCLTest 
[==========] Running 4 tests from 1 test case.
[----------] Global test environment set-up.
[----------] 4 tests from OpenCLCorrectnessTest
[ RUN      ] OpenCLCorrectnessTest.convOps
[       OK ] OpenCLCorrectnessTest.convOps (49 ms)
[ RUN      ] OpenCLCorrectnessTest.inferMixedNet
[       OK ] OpenCLCorrectnessTest.inferMixedNet (20 ms)
[ RUN      ] OpenCLCorrectnessTest.softmaxGradTest
[       OK ] OpenCLCorrectnessTest.softmaxGradTest (105 ms)
[ RUN      ] OpenCLCorrectnessTest.tanhConcatTest
[       OK ] OpenCLCorrectnessTest.tanhConcatTest (10 ms)
[----------] 4 tests from OpenCLCorrectnessTest (184 ms total)

[----------] Global test environment tear-down
[==========] 4 tests from 1 test case ran. (185 ms total)
[  PASSED  ] 4 tests.
```

@ffk0716 Which `cmake` version are you using? I'm using `3.12.1`.

**Comment by ffk0716** [comment]
I am using cmake version 3.14.2

File: lib/Backends/OpenCL/CMakeLists.txt (hunk 1 above)
Comment (rdzhabarov) [reply lib/Backends/OpenCL/CMakeLists.txt]: hehe, no idea, but @opti-mix you introduced that in
https://github.com/pytorch/glow/commit/2b9abbc012437004e6b9766ded2b400903598a07

**Comment by rdzhabarov** [comment]
Pure CMake changes should not affect any internal components anyhow (we should be good from that point of view).

File: lib/Backends/OpenCL/CMakeLists.txt (hunk 1 above)
Comment (opti-mix) [reply lib/Backends/OpenCL/CMakeLists.txt]: Ha-ha. So, it was me??? :-D I'm almost sure it is a copy&paste error.

**Comment by ffk0716** [comment]
Hello?

**Comment by SplitInfinity** [comment]
@ffk0716 

I am not able to reproduce this issue using `cmake v 3.12.1` on macOS Mojave 10.14.4. What are the details of your setup?

**Comment by ffk0716** [comment]
Did you enable OpenCL backend?
Which command generates ${CMAKE_BINARY_DIR}/glow/OpenCL in your case?

In my case:
```
mkdir build
cd build
cmake .. -DGLOW_WITH_OPENCL=ON
ls glow/OpenCL
ls: glow/OpenCL: No such file or directory
make
[ 32%] Generating ../../../glow/OpenCL/kernels_fwd_quantized_conv.cl.inc
Could not open output file: No such file or directory

```

**Comment by rdzhabarov** [comment]
@ffk0716 

Could you try ninja generator in cmake? (I can run cmake/ninja with cmake version 3.14.3 with no issues).
I do not think we were able to use GNU make for long time in Glow.

**Comment by SplitInfinity** [comment]
I was just about to comment - I was also able to build just fine with `Ninja`.

**Comment by ffk0716** [comment]
ninja works fine.
It seems gnu make does not mkdir automatically.
But ninja does.

**Comment by rdzhabarov** [comment]
@ffk0716 
Thanks for digging into the problem!

I'd suggest to simplify a bit and just  run `file(MAKE_DIRECTORY ${OpenCL_DIR})` and remove custom target.

see: https://github.com/pytorch/glow/blob/e02b8732729bed843401398c1ab1822c9b623c09/lib/Backends/CPU/CMakeLists.txt#L27 for example.

**Comment by ffk0716** [comment]
`file()` also works in this case.

But if we delete folder and remake.
comstom_target works, but `file()` not.

**Comment by SplitInfinity** [comment]
No more outstanding questions or issues.

**Pull request closed without merging**